File: Source/OldCode/Archive/2005 TVA Code Library/Version/2.0/Local/Source/Tva.Phasors/PhasorDefinitionBase.vb
'*******************************************************************************************************
'  PhasorDefinitionBase.vb - Phasor value definition base class
'  Copyright © 2005 - TVA, all rights reserved - Gbtc
'
'  Build Environment: VB.NET, Visual Studio 2005
'  Primary Developer: J. Ritchie Carroll, Operations Data Architecture [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  -----------------------------------------------------------------------------------------------------
'  02/18/2005 - J. Ritchie Carroll
'       Initial version of source generated
'
'*******************************************************************************************************

Imports System.Runtime.Serialization

''' <summary>This class represents the common implementation of the protocol independent definition of a phasor value.</summary>
<CLSCompliant(False), Serializable()> _
Public MustInherit Class PhasorDefinitionBase

    Inherits ChannelDefinitionBase
    Implements IPhasorDefinition

    Private m_type As PhasorType
    Private m_voltageReference As IPhasorDefinition

    Protected Sub New()
    End Sub

    Protected Sub New(ByVal info As SerializationInfo, ByVal context As StreamingContext)

        MyBase.New(info, context)

        ' Deserialize phasor definition
        m_type = info.GetValue("type", GetType(PhasorType))
        m_voltageReference = info.GetValue("voltageReference", GetType(IPhasorDefinition))

    End Sub

    Protected Sub New(ByVal parent As IConfigurationCell)

        MyBase.New(parent)

        m_type = PhasorType.Voltage
        m_voltageReference = Me

    End Sub

    Protected Sub New(ByVal parent As IConfigurationCell, ByVal index As Int32, ByVal label As String, ByVal scale As Int32, ByVal offset As Single, ByVal type As PhasorType, ByVal voltageReference As IPhasorDefinition)

        MyBase.New(parent, index, label, scale, offset)

        m_type = type

        If type = PhasorType.Voltage Then
            m_voltageReference = Me
        Else
            m_voltageReference = voltageReference
        End If

    End Sub

    Protected Sub New(ByVal parent As IConfigurationCell, ByVal binaryImage As Byte(), ByVal startIndex As Int32)

        MyBase.New(parent, binaryImage, startIndex)

    End Sub

    ' Derived classes are expected to expose a Public Sub New(ByVal phasorDefinition As IPhasorDefinition)
    Protected Sub New(ByVal phasorDefinition As IPhasorDefinition)

        MyClass.New(phasorDefinition.Parent, phasorDefinition.Index, phasorDefinition.Label, _
            phasorDefinition.ScalingFactor, phasorDefinition.Offset, phasorDefinition.Type, phasorDefinition.VoltageReference)

    End Sub

    Public Overrides ReadOnly Property DataFormat() As DataFormat
        Get
            Return Parent.PhasorDataFormat
        End Get
    End Property

    Public Overridable ReadOnly Property CoordinateFormat() As CoordinateFormat Implements IPhasorDefinition.CoordinateFormat
        Get
            Return Parent.PhasorCoordinateFormat
        End Get
    End Property

    Public Overridable Property [Type]() As PhasorType Implements IPhasorDefinition.Type
        Get
            Return m_type
        End Get
        Set(ByVal value As PhasorType)
            m_type = value
        End Set
    End Property

    Public Overridable Property VoltageReference() As IPhasorDefinition Implements IPhasorDefinition.VoltageReference
        Get
            Return m_voltageReference
        End Get
        Set(ByVal value As IPhasorDefinition)
            If m_type = PhasorType.Voltage Then
                If Not value Is Me Then
                    Throw New NotImplementedException("Voltage phasors do not have a voltage reference")
                End If
            Else
                m_voltageReference = value
            End If
        End Set
    End Property

    Public Overrides Sub GetObjectData(ByVal info As System.Runtime.Serialization.SerializationInfo, ByVal context As System.Runtime.Serialization.StreamingContext)

        MyBase.GetObjectData(info, context)

        ' Serialize phasor definition
        info.AddValue("type", m_type, GetType(PhasorType))
        info.AddValue("voltageReference", m_voltageReference, GetType(IPhasorDefinition))

    End Sub

    Public Overrides ReadOnly Property Attributes() As System.Collections.Generic.Dictionary(Of String, String)
        Get
            MyBase.Attributes.Add("Phasor Type", [Type] & ": " & [Enum].GetName(GetType(PhasorType), [Type]))
            Return MyBase.Attributes
        End Get
    End Property

End Class

File: Source/Libraries/GSF.PhasorProtocols/PhasorDefinitionBase.cs
//******************************************************************************************************
//  PhasorDefinitionBase.cs - Gbtc
//
//  Copyright © 2012, Grid Protection Alliance.  All Rights Reserved.
//
//  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
//  the NOTICE file distributed with this work for additional information regarding copyright ownership.
//  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may
//  not use this file except in compliance with the License. You may obtain a copy of the License at:
//
//      http://www.opensource.org/licenses/MIT
//
//  Unless agreed to in writing, the subject software distributed under the License is distributed on an
//  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
//  License for the specific language governing permissions and limitations.
//
//  Code Modification History:
//  ----------------------------------------------------------------------------------------------------
//  02/18/2005 - J. Ritchie Carroll
//       Generated original version of source code.
//  12/17/2012 - Starlynn Danyelle Gilliam
//       Modified Header.
//
//******************************************************************************************************

using System;
using System.Collections.Generic;
using System.Runtime.Serialization;
using GSF.Units.EE;

namespace GSF.PhasorProtocols
{
    /// <summary>
    /// Represents the common implementation of the protocol independent definition of a <see cref="IPhasorValue"/>.
    /// </summary>
    [Serializable]
    public abstract class PhasorDefinitionBase : ChannelDefinitionBase, IPhasorDefinition
    {
        #region [ Members ]

        // Fields
        private PhasorType m_type;
        private IPhasorDefinition m_voltageReference;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="PhasorDefinitionBase"/>.
        /// </summary>
        /// <param name="parent">The <see cref="IConfigurationCell"/> parent of this <see cref="PhasorDefinitionBase"/>.</param>
        protected PhasorDefinitionBase(IConfigurationCell parent)
            : base(parent)
        {
        }

        /// <summary>
        /// Creates a new <see cref="PhasorDefinitionBase"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The <see cref="IConfigurationCell"/> parent of this <see cref="PhasorDefinitionBase"/>.</param>
        /// <param name="label">The label of this <see cref="PhasorDefinitionBase"/>.</param>
        /// <param name="scale">The integer scaling value of this <see cref="PhasorDefinitionBase"/>.</param>
        /// <param name="offset">The offset of this <see cref="PhasorDefinitionBase"/>.</param>
        /// <param name="type">The <see cref="PhasorType"/> of this <see cref="PhasorDefinitionBase"/>.</param>
        /// <param name="voltageReference">The associated <see cref="IPhasorDefinition"/> that represents the voltage reference (if any).</param>
        protected PhasorDefinitionBase(IConfigurationCell parent, string label, uint scale, double offset, PhasorType type, IPhasorDefinition voltageReference)
            : base(parent, label, scale, offset)
        {
            m_type = type;

            if (type == PhasorType.Voltage)
                m_voltageReference = this;
            else
                m_voltageReference = voltageReference;
        }

        /// <summary>
        /// Creates a new <see cref="PhasorDefinitionBase"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected PhasorDefinitionBase(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize phasor definition
            m_type = (PhasorType)info.GetValue("type", typeof(PhasorType));
            m_voltageReference = (IPhasorDefinition)info.GetValue("voltageReference", typeof(IPhasorDefinition));
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets the <see cref="GSF.PhasorProtocols.DataFormat"/> of this <see cref="PhasorDefinitionBase"/>.
        /// </summary>
        public override DataFormat DataFormat
        {
            get
            {
                return Parent.PhasorDataFormat;
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="GSF.PhasorProtocols.CoordinateFormat"/> of this <see cref="PhasorDefinitionBase"/>.
        /// </summary>
        public virtual CoordinateFormat CoordinateFormat
        {
            get
            {
                return Parent.PhasorCoordinateFormat;
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="GSF.PhasorProtocols.AngleFormat"/> of this <see cref="PhasorDefinitionBase"/>.
        /// </summary>
        public virtual AngleFormat AngleFormat
        {
            get
            {
                return Parent.PhasorAngleFormat;
            }
        }

        /// <summary>
        /// Gets or sets the <see cref="PhasorType"/> of this <see cref="PhasorDefinitionBase"/>.
        /// </summary>
        public virtual PhasorType PhasorType
        {
            get
            {
                return m_type;
            }
            set
            {
                m_type = value;
            }
        }

        /// <summary>
        /// Gets or sets the associated <see cref="IPhasorDefinition"/> that represents the voltage reference (if any).
        /// </summary>
        /// <remarks>
        /// This only applies to current phasors.
        /// </remarks>
        public virtual IPhasorDefinition VoltageReference
        {
            get
            {
                return m_voltageReference;
            }
            set
            {
                if (m_type == PhasorType.Voltage)
                    m_voltageReference = this;
                else
                    m_voltageReference = value;
            }
        }

        /// <summary>
        /// Gets a <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for this <see cref="PhasorDefinitionBase"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("Phasor Type", (int)PhasorType + ": " + PhasorType);

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Gets the string representation of this <see cref="PhasorDefinitionBase"/>.
        /// </summary>
        /// <returns>String representation of this <see cref="PhasorDefinitionBase"/>.</returns>
        public override string ToString()
        {
            return (PhasorType == PhasorType.Current ? "I: " : "V: ") + Label;
        }

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize phasor definition
            info.AddValue("type", m_type, typeof(PhasorType));
            info.AddValue("voltageReference", m_voltageReference, typeof(IPhasorDefinition));
        }

        #endregion
    }
}